vyper/venom/passes/memmerging.py

```diff
@@ -0,0 +1,336 @@
+                load_inst = self.dfg.get_producing_instruction(var)
+                assert load_inst is not None  # help mypy
+                n_copy = _Copy(dst.value, src_ptr, 32, [inst, load_inst])
+
+                if self._write_after_write_hazard(n_copy):
+                    _barrier()
+                # check if the new copy does not overwrites existing data
+                if not allow_dst_overlaps_src and self._read_after_write_hazard(n_copy):
+                    _barrier()
+                self._add_copy(n_copy)
+
+            elif inst.opcode == copy_opcode:
+                if not all(isinstance(op, IRLiteral) for op in inst.operands):
+                    _barrier()
+                    continue
+
+                length, src, dst = inst.operands
+                n_copy = _Copy(dst.value, src.value, length.value, [inst])
+
+                if self._write_after_write_hazard(n_copy):
+                    _barrier()
+                # check if the new copy does not overwrites existing data
+                if not allow_dst_overlaps_src and self._read_after_write_hazard(n_copy):
+                    _barrier()
+                self._add_copy(n_copy)
+
+            elif _volatile_memory(inst):
+                _barrier()
+
+        _barrier()
+        bb.clear_dead_instructions()
+
+    # optimize memzeroing operations
+    def _optimize_memzero(self, bb: IRBasicBlock):
+        for copy in self._copies:
+            inst = copy.insts[0]
+            if copy.length == 32:
+                inst.opcode = "mstore"
+                inst.operands = [IRLiteral(0), IRLiteral(copy.dst)]
+            else:
+                index = bb.instructions.index(inst)
+                calldatasize = bb.parent.get_next_variable()
+                bb.insert_instruction(IRInstruction("calldatasize", [], output=calldatasize), index)
+
+                inst.output = None
+                inst.opcode = "calldatacopy"
+                inst.operands = [IRLiteral(copy.length), calldatasize, IRLiteral(copy.dst)]
+
+            for inst in copy.insts[1:]:
+                bb.mark_for_removal(inst)
+
+        self._copies.clear()
+        self._loads.clear()
+
+    def _handle_bb_memzero(self, bb: IRBasicBlock):
+        self._loads = {}
+        self._copies = []
+
+        def _barrier():
+            self._optimize_memzero(bb)
+
+        # copy in necessary because there is a possibility
+        # of insertion in optimizations
+        for inst in bb.instructions.copy():
+            if inst.opcode == "mstore":
+                val = inst.operands[0]
+                dst = inst.operands[1]
+                is_zero_literal = isinstance(val, IRLiteral) and val.value == 0
+                if not (isinstance(dst, IRLiteral) and is_zero_literal):
+                    _barrier()
+                    continue
+                n_copy = _Copy(dst.value, dst.value, 32, [inst])
+                if self._write_after_write_hazard(n_copy):
```

Comment:
Is this barier needed here since even tho it could be triggered this pass only handles writing zeros it should not cause the problem?

Reply:
Same thing is done in other case of zeroing (calldatacopy case)

Reply:
yea, that barrier is unreachable (since for memzero, overwrites(dst) implies can_merge()==True)